### PR TITLE
Foundation of New Product Collection Block

### LINKED ISSUE #9267: Foundations of the Product Collection
### Description
Provide foundations to the Product Collection block, by reusing Query Loop and Post Template code (copy-pasting/forking) and adjusting it if necessary.

### Outcomes
New blocks:
- Product Collection
- Product Template

### Goals - acceptance criteria
- Block may have hardcoded basic query,
- Block displays the product grid in the editor and frontend,
- Product Elements can be added/removed to the Product Template and basic operations can be made (reordering, changing Product Element settings),
- Product Collection has updated icon and description
- There are pagination and "No results" blocks included
- At this stage block may have bugs, but the happy path should work

### PR DESCRIPTION
**Pull request by imanish003**
This PR introduces the foundations for a new Product Collection block. It includes the creation of two new blocks:
1. Product Collection
2. Product Template

The Product Collection block will replace the existing Products (Beta) block. When the Product Collection block is added to the Editor, it will display all products in a grid view. On the frontend, all products should be displayed, and pagination should function as expected.

Fixes #9267

### Testing

1. Create a new post 
2. Add the Product Collection block to the post.
3. Verify that the UI of the Product Collection block matches that of the existing Products Beta block in both the editor and the frontend.
4. Test the pagination functionality on the frontend to ensure it works as expected.

Here is a quick demo:

https://user-images.githubusercontent.com/16707866/236784525-9bb93c6d-e538-40d3-9524-4d71f91d4431.mov

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Foundation of New `Product Collection` Block

### CODE REVIEW

**Comment by roykho** [comment]
Some of the frontend JS files are not loaded correctly:


When I put the `Products (beta)` block and `Products Collection` block side by side, I can see different loading orders of the products. I am not sure if this is an issue or not. Looking at your example, they're the same. But on mine they are not.

**Comment by imanish003** [comment]
Hi @roykho, thanks for the feedback 🙇 

> Some of the frontend JS files are not loaded correctly:

I couldn't replicate this issue. Could you please try running `rm -rf build` to remove the build directory, and then execute `npm start`? This should resolve the error. 🙂

> When I put the Products (beta) block and Products Collection block side by side, I can see different loading orders of the products. I am not sure if this is an issue or not. Looking at your example, they're the same. But on mine they are not.

Apologies for the confusion. In fact, I manually changed the order for the Products beta block for comparison purposes. I selected `Newest` in the dropdown. Currently, the order is hardcoded in the block.json file [here](https://github.com/woocommerce/woocommerce-blocks/blob/94a75c67b31c1afbea4dc7ec3c597dd5cd60c534/assets/js/blocks/products-collection/block.json#L22-L23). However, we plan to add a dropdown in the sidebar settings, allowing users to modify the order.

image

I've just pushed [aace308](https://github.com/woocommerce/woocommerce-blocks/pull/9352/commits/aace3080edfc85774dd75d53a4bd524fa618a310) to align the default ordering with the Products beta block. Now, both blocks will display the same order by default. 🙂

File: assets/js/blocks/product-template/index.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+const blockConfig: BlockConfiguration = {
+	...metadata,
+	icon,
+	edit,
+	save,
+};
+
+if ( isExperimentalBuild() ) {
+	registerBlockType( metadata.name, blockConfig );
+}
```
Comment (kmanijak) [inline assets/js/blocks/product-template/index.tsx]: You can use `registerExperimentalBlockType` function available in:

`import { registerExperimentalBlockType } from '@woocommerce/block-settings';`

Although it's basically doing the same thing, so leaving the decision up to you.

**Review by kmanijak — COMMENTED** [review]
`assets/js/blocks/product-template/types.ts` is an empty file, so I suggest removing it for now.

**Comment by roykho** [comment]
Even after removing the build directory and re-running `npm start` I am still seeing the same missing `JS` files error in console.

**Comment by kmanijak** [comment]
> Even after removing the build directory and re-running npm start I am still seeing the same missing JS files error in console.

@imanish003, I think this piece of code is required to prevent the `*-frontend.js` file to be fetched for:

```
	/**
	 * Get the frontend script handle for this block type.
	 *
	 * @param string $key Data to get, or default to everything.
	 */
	protected function get_block_type_script( $key = null ) {
		return null;
	}
```

I remember adding it to the block's class for the same reason. Here are examples:
- [ProductResultsCount.php](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTypes/ProductResultsCount.php#L24)
- [CatalogSorting.php](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTypes/CatalogSorting.php#L54)

**Comment by imanish003** [comment]
@kmanijak Oh yes, thank you 🙏🏻! I believe implementing your suggested solution will resolve the issue.

File: assets/js/blocks/product-template/index.tsx (hunk 1 above)
Comment (imanish003) [reply assets/js/blocks/product-template/index.tsx]: I'm going to go ahead and use registerExperimentalBlockType to make the code a bit cleaner. If there's an abstraction available, I think it's better to use it. Thanks Karol.

**Comment by imanish003** [comment]
@roykho 
> Even after removing the build directory and re-running npm start, I am still encountering the same missing JS files error in the console.

Oh, my bad! I was checking on the Editor side, but I am able to reproduce it on the frontend side. I was able to fix it by following Karol's suggestion in the comment above.

@kmanijak, I have also implemented the other minor improvements you suggested. Could you please take another quick look? 🙏🏻

File: assets/js/blocks/product-template/index.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/**
+ * External dependencies
+ */
+import { BlockConfiguration, registerBlockType } from '@wordpress/blocks';
```
Comment (roykho) [inline assets/js/blocks/product-template/index.tsx]: Perhaps remove this `registerBlockType` since it is not used.

**Review by roykho — APPROVED** [review]
Just a small comment but not a blocker. Otherwise LGTM. The frontend JS errors are now gone. Good job!

**Review by kmanijak — CHANGES_REQUESTED** [review]
I just spotted one tiny, but significant thing 😅 

Block is supposed to be named Product Collection, not Products Collection (singular, not plural).

From ref: pdnLyh-3uF-p2
> Rename from Products to Product Collection

Let's unify it now, since it's a kick-off PR, so that would mean:
- changing the block name and its references
- changing the directory in which the block is stored.

**Comment by imanish003** [comment]
Hi @kmanijak, I actually had the intention of confirming this with @shaunandrews, but I forgot to do so. 🤦‍♂️

@shaunandrews, just to confirm, we want to name the new block as `Product Collection` instead of `Products Collection`, correct? 🙂

**Comment by imanish003** [comment]
Hey @kmanijak, thank you for pointing this out. I have made the necessary changes in [f3453cc](https://github.com/woocommerce/woocommerce-blocks/pull/9352/commits/f3453ccdc15014b85c93988bb426edb906506010). Could you please take a quick look?

**Review by kmanijak — APPROVED** [review]
Thanks for all the changes! 💪 Please also change Products Collection to Product Collection in the PR title and changelog. 🙏 

One last thing I found is that Product Template should not be available in the inserter. But anyway, I'm pre-approving!

**Comment by imanish003** [comment]
Hey @kmanijak, good catch! 🙌🏻

I have made the changes to the title and description of this PR. Additionally, after the changes made in commit [9bc39b2](https://github.com/woocommerce/woocommerce-blocks/pull/9352/commits/9bc39b22482f4f6f9025d9d30133140c971422de), the Product Template block will no longer be available in the inserter. 🤝

**Pull request merged**